scripts/pylib/twister/twisterlib/handlers.py

```diff
@@ -546,15 +588,36 @@ def handle(self):
         if hardware.flash_with_test:
             flash_timeout += self.timeout
 
+        harness_name = self.instance.testsuite.harness.capitalize()
+        harness_import = HarnessImporter(harness_name)
+        harness = harness_import.instance
+        harness.configure(self.instance)
+        halt_monitor_evt = threading.Event()
+
+        # Create serial device but don't open it yet
+        ser = serial.Serial(
+            baudrate=hardware.baud,
+            parity=serial.PARITY_NONE,
+            stopbits=serial.STOPBITS_ONE,
+            bytesize=serial.EIGHTBITS,
+            timeout=max(flash_timeout, self.timeout)  # the worst case of no serial input
+        )
+
+        flash_before = hardware.flash_before
+
+        if flash_before:
+            t = threading.Thread(target=self.monitor_serial, daemon=True,
+                    args=(ser, halt_monitor_evt, harness))
+            start_time = time.time()
+            t.start()
+            flash_error = self.run_flash(command, flash_timeout, halt_monitor_evt)
+            if post_flash_script:
+                self.run_custom_script(post_flash_script, 30)
+
+        # Connect to device
         try:
-            ser = serial.Serial(
-                serial_device,
-                baudrate=hardware.baud,
-                parity=serial.PARITY_NONE,
-                stopbits=serial.STOPBITS_ONE,
-                bytesize=serial.EIGHTBITS,
-                timeout=max(flash_timeout, self.timeout)  # the worst case of no serial input
-            )
+            ser.port = serial_device
+            ser.open()
```

Comment:
In case of a USB UART that is disconnected from the host PC during flash, and reconnects after flash, this may fail.

```python
USB_CDC_ACM_CONNECTION_TIMEOUT_S = 10
usb_cdc_acm_connection_time = usb_cdc_acm_start_time = time.time()

while usb_cdc_acm_connection_time < usb_cdc_acm_start_time + USB_CDC_ACM_CONNECTION_TIMEOUT_S:
    usb_cdc_acm_connection_time = time.time()
    try: 
        ser.open()
    except serial.SerialException:
        time.sleep(0.05)
        continue

if not ser.is_open():
    # handle error
```

Reply:
Super uncomfortable that monitor serial is a python thread lmao